teps/0115-tekton-catalog-git-based-versioning.md

```diff
@@ -0,0 +1,398 @@
+
+### Git-Based Versioning
+
+#### Versioning via Tags
+
+Versioning of resources will be done through creating tags in the Catalogs. The versions in the Tekton Hub should
+continue to use the simple version of [semver][semver]: `{major}.{minor}`. We propose that the tags use the semantic
+versioning model - `v<major-version>.<minor-version>`. Minor versions are made to signal backwards-compatible changes,
+while major versions are made to signal backwards-incompatible changes. As such, the tags in the catalogs will map to
+versions in the Hub e.g. tag "v0.7" in a Catalog maps to version "0.7" in the Hub.
+
+#### One Version per Catalog
+
+Note that version tags apply to the whole repository. Therefore, we propose that Catalogs should contain one or
+multiple resource that have the same versioning. Instead of the centralized Catalog from the community, we will
+encourage users to create Catalogs in their own GitHub organizations and share them through the Hub.
+
+##### Examples
+
+###### Buildpacks
+
+[Buildpacks][buildpacks] has a Catalog with a single `Task`. The team added the same `Task` to the Tekton Community 
+Catalog - [Buildpacks][buildpacks-catalog]. They add subdirectories in the directory for that `Task` in the Tekton 
+Community Catalog to add new versions of the `Task`. With the above solution, the Buildpacks team will host their own
+Catalog without having to copy it to Tekton Catalog for directory-based versioning. Instead, they would make releases
+in their own [Buildpacks][buildpacks] Catalog and the Hub will pick up the versions based on the tags in the repository.
+
+###### eBay
+
+[eBay][ebay] shared a repository with a single `Task` for sending Slack messages. The team could modify the repository
+to meet the new organization contract, and share this `Task` in the Hub without adding it to a centralized Catalog. eBay
+can version the resources using tags, which would be picked up by the Tekton Hub.
+
+#### Migration
+
+The Hub should support both the old and new versioning to provide backwards compatibility. The [config.yaml][cy] will
+take a `versioning` field under `catalogs`, which will default to `"directory"` and users can set it to `"git"`. Taking
+the examples above, the entries in [config.yaml][cy] could be:
+
+```yaml
+catalogs:
+
+- name: tekton
+  org: tektoncd
+  type: community
+  provider: github
+  url: https://github.com/tektoncd/catalog
+
+- name: buildpacks
+  org: buildpacks
+  type: community
+  provider: github
+  url: https://github.com/buildpacks/tekton-integration
+  versioning: git
+
+- name: slack-notify
+  org: ebay
+  type: community
+  provider: github
+  url: https://github.com/eBay/tekton-slack-notify
+  versioning: git
+ ```
+
+The existing Tekton Catalog - `tektoncd/catalog` - will remain as it is using the old contract. It will be read-only
+and will encourage owners of the resources to move them to their own repositories in their own GitHub organizations.
```

Comment:
I think we need to expand on how this move will take happen.
Will we provide a github template repo which includes CI resources? How will that look like? 
If we use prow we could use some gitops tool to deploy distributed configs to the cluster. 
If we use Tekton we could use git remote resolution. 
What about repo ownership? If a new repo is added, what will be the process to setup owners for it? 
I think everything in the `tektoncd` org should be owned by the catalog maintainers team, as this approach makes it easier to host tasks outside of `tektoncd`. 
We will probably need to adapt the contributor ladder for catalog contributors as they will be spread across multiple repos. Would it make sense to use a different GitHub org `tektoncd-catalog` (https://github.com/tektoncd-catalog is taken, I hope by someone at Tekton).

Reply:
I think @wlynch has tekoncd-catalog ? I think we can have the catalog maintainers own the org.

template repo that provides a template for task CI sounds good.

Reply:
Yeah I parked that org a few weeks back when I saw this proposal just in case. Consider it owned by tektoncd.

Reply:
Thanks for reserving that org for us @wlynch!

@afrittoli these questions and suggestions have been very useful in adding more details to the TEP, please take a look 😁